src/server/string_family.cc

```diff
@@ -356,6 +359,92 @@ OpResult<void> SetGeneric(ConnectionContext* cntx, const SetCmd::SetParams& spar
   return cntx->transaction->ScheduleSingleHop(std::move(cb));
 }
 
+OpResult<array<int64_t, 5>> OpThrottle(const OpArgs& op_args, string_view key, int64_t max_burst,
+                                       int64_t count, int64_t period_s, int64_t quantity) {
+  using namespace chrono_literals;
+
+  auto& db_slice = op_args.shard->db_slice();
+
+  const int64_t limit = max_burst + 1;
+  const int64_t emission_interval_ms = period_s * 1000 / count;
+  const int64_t delay_variation_tolerance_ms = emission_interval_ms * limit;
+
+  if (emission_interval_ms == 0) {
+    return OpStatus::OUT_OF_RANGE;
+  }
+
+  int64_t remaining = 0;
+  int64_t reset_after_ms = -1000;
+  int64_t retry_after_ms = -1000;
+
+  const int64_t increment_ms = emission_interval_ms * quantity;
+
+  auto [it, expire_it] = db_slice.FindExt(op_args.db_cntx, key);
+  const int64_t now_ms = op_args.db_cntx.time_now_ms;
+
+  int64_t tat_ms = now_ms;
+  if (IsValid(it)) {
+    if (it->second.ObjType() != OBJ_STRING) {
+      return OpStatus::WRONG_TYPE;
+    }
+
+    auto opt_prev = it->second.TryGetInt();
+    if (!opt_prev) {
+      return OpStatus::INVALID_VALUE;
+    }
+    tat_ms = *opt_prev;
+  }
+  const int64_t new_tat_ms = max(tat_ms, now_ms) + increment_ms;
+
+  const int64_t allow_at_ms = new_tat_ms - delay_variation_tolerance_ms;
+  const int64_t diff_ms = now_ms - allow_at_ms;
+
+  const bool limited = diff_ms < 0;
+  int64_t ttl_ms;
+  if (limited) {
+    if (increment_ms <= delay_variation_tolerance_ms) {
+      retry_after_ms = -diff_ms;
+    }
+    ttl_ms = tat_ms - now_ms;
+  } else {
+    ttl_ms = new_tat_ms - now_ms;
+
+    if (IsValid(it)) {
+      db_slice.PreUpdate(op_args.db_cntx.db_index, it);
+      it->second.SetInt(new_tat_ms);
+      db_slice.PostUpdate(op_args.db_cntx.db_index, it, key);
+    } else {
+      CompactObj cobj;
+      cobj.SetInt(new_tat_ms);
+
+      // AddNew calls PostUpdate inside.
+      try {
+        it = db_slice.AddNew(op_args.db_cntx, key, std::move(cobj), 0);
+      } catch (bad_alloc&) {
+        return OpStatus::OUT_OF_MEMORY;
+      }
+    }
+  }
+
+  const int64_t next_ms = delay_variation_tolerance_ms - ttl_ms;
+  if (next_ms > -emission_interval_ms) {
+    remaining = next_ms / emission_interval_ms;
+  }
+  reset_after_ms = ttl_ms;
+
+  int64_t retry_after_s = retry_after_ms / 1000;
```

Comment:
I suspect that your original intent was to convert  `1500` to 2, right?
because this is not what this code does.

Reply:
the right math is: `(x + k -1)/k`, or in this case `(x + 999) / 1000`

Reply:
I've checked the original does this too. It's not "divide with ceil".

Reply:
can you please revisit it again?
I am looking at this code: https://github.com/brandur/redis-cell/blob/master/src/lib.rs#L69-L76
not a rust developer but does not it means that duration is composed from `(seconds, milliseconds)` and milliseconds are essentially the remainder, i.e. always in the range `0-999` ? in other words the original code checks if the remainder exists?

Reply:
Ok, looked further - seems you are correct and based on https://docs.rs/time/0.1.40/time/struct.Duration.html#method.num_milliseconds 
the function returns the total number of ms, so you are factually correct.
Having said that, I do think that the author's intent was different and this is a bug in his code, based on the comment he wrote:

````rust
  // If either time had a partial component, but it up to the next full
  // second because otherwise a fast-paced caller could try again too
  // early.
```

To summarize - whatever you decide it's fine but the this code does not make much sense :)

Reply:
🤷‍♀️